includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -1664,16 +1668,18 @@ public function capture_charge( $order ) {
 				WC_Payments_Explicit_Price_Formatter::get_explicit_price( wc_price( $amount, [ 'currency' => $currency ] ), $order )
 			);
 			$order->add_order_note( $note );
+			$http_code = 500;
```

Comment:
I'm not sure if `500` is the appropriate code in case of capture amount failure.

Reply:
I think it should be `400`, which is similar to the main issue here.

It looks like that right after the current line 1660 should have `$http_code = 400` too;

Reply:
@htdat You mean the errors on 1671 and 1660 are not server errors but client errors (invalid input data)? I do not have enough context of all possible errors possible at these locations, so I relied on the unit tests for my understanding.

Eg. in `test_capture_charge_failure` and `test_capture_charge_failure_non_usd`, the error seems to be caused by the payment status remaining the same ("requires_charge") between order creation and capturing of intent. My assumption was that the payment status would not change if the upstream Stripe API somehow failed on calling `capture_intention`.

In other tests — `test_capture_charge_api_failure`, `test_capture_charge_api_failure_non_usd`, and `test_capture_charge_expired` — a mock exception is explicitly thrown with HTTP code 500.

Btw, I missed setting `$http_code` on line 1661. Thanks for catching that!

Reply:
Actually, I now remember why I didn't set `$http_code` on line 1661. That's because it's only possible to enter that `elseif` block when `$error_message` is non-empty:

https://github.com/Automattic/woocommerce-payments/blob/f35dd3dbe6ec04e2d5fdc03a7483517ca5f70573/includes/class-wc-payment-gateway-wcpay.php#L1606-L1622

On lines 1609 and 1621, we set `$http_code` along with `$error_message`. In other words,  `$http_code` will have been set if `$error_message` is set. So, there is no need to explicitly set it again on 1661. Doing so would override the upstream API response code.

However, it was necessary to set `$http_code` on line 1671 to handle cases as described in `test_capture_charge_failure` and `test_capture_charge_failure_non_usd`.

Reply:
> Actually, I now remember why I didn't set `$http_code` on line 1661. That's because it's only possible to enter that `elseif` block when `$error_message` is non-empty
> [...]

Ah, that makes sense for me now. If so, nitpick: is it worth adding a comment here?

Reply:
**Regarding `400` or `500` status:** 

> Eg. in test_capture_charge_failure and test_capture_charge_failure_non_usd, the error seems to be caused by the payment status remaining the same ("requires_charge") between order creation and capturing of intent... 

I think this test means if an error happens when we try to capture a charge, Stripe will keep PaymentIntent's status as `requires_charge`. 

> In other tests — `test_capture_charge_api_failure`, `test_capture_charge_api_failure_non_usd`, and `test_capture_charge_expired` — a mock exception is explicitly thrown with HTTP code 500.

All $http_code places in those tests are added in your PR so I don't think it's reasonable to use `$http_code` as a judgment here.

---- 

I also looked at this PR https://github.com/Automattic/woocommerce-payments/pull/106, where this error message `A capture of %1$s <strong>failed</strong> to complete.` was added, and it's not clear when this can happen. **Therefore, I agreed that this is a grey area and I don't have a strong opinion on this anymore.** Let's wait for other reviewers' opinions on this issue.

Reply:
Thanks for your thoughts, @htdat. In that case, pinging @kalessil and @marcinbot for their thoughts on this particular thread.